CONTRIBUTING.md

```diff
@@ -78,6 +78,123 @@ available in bootstrap classloader.
+using library-specific monitoring APIs. For example, when instrumenting an RPC library,
+the instrumentation will use some library-specific functionality to listen to events such
+as the start and end of a request and will execute code to start and end spans in these
+listeners. Many of these libraries will provide interception type APIs such as the gRPC
+`ClientInterceptor` or servlet's `Filter`. Others will provide a Java interface whose methods
+correspond to a request, and instrumentation can define an implementation which delegates
+to the standard, wrapping methods with the logic to manage spans. Users will add code to their
+apps that initialize the classes provided by manual instrumentation libraries and the libraries
+can be found inside the user's app itself.
+
+Some libraries will have no way of intercepting requests because they only expose static APIs
+and no interception hooks. For these libraries it is not possible to create manual
+instrumentation.
+
+**Auto Instrumentation**: This is logic that is similar to manual instrumentation, but instead
+of a user initializing classes themselves, a Java agent automatically initializes them during
+class loading by manipulating byte code. This allows a user to develop their apps without thinking
+about instrumentation and get it "for free". Often, the auto instrumentation will generate bytecode
+that is more or less identical to what a user would have written themselves in their app.
+
+In addition to automatically initializing manual instrumentation, auto instrumentation can be used
+for libraries where manual instrumentation is not possible, such as `URLConnection`, because it can
+intercept even the JDK's classes. Such libraries will not have manual instrumentation but will have
+auto instrumentation.
+
+#### Folder Structure
+
+When writing new instrumentation, create a new subfolder of `instrumentation` to correspond to the
+instrumented library and the version being targeted. Ideally an old version of the library is
+targeted in a way that the instrumentation applies to a large range of versions, but this may be
+restricted by the interception APIs provided by the library.
+
+Within the subfolder, create three folders `library` (skip if manual instrumentation is not possible),
+`agent`, and `testing`.
+
+For example, if we are targeting an RPC framework `yarpc` at version `1.0` we would have a tree like
+
+```
+instrumentation ->
+    ...
+    yarpc-1.0 ->
+        agent
+            agent.gradle
+        library
+            library.gradle
+        testing
+            testing.gradle
+```
+
+and in the top level `settings.gradle`
+
+```groovy
+
+include 'instrumentation:yarpc-1.0:agent'
+include 'instrumentation:yarpc-1.0:library'
+include 'instrumentation:yarpc-1.0:testing'
+```
+
+#### Writing manual instrumentation
+
+Begin by writing the instrumentation for the library in `library`. This generally involves defining a
```

Comment:
This doc is a very good starting point. But we have to write more extensive guide with code examples and, probably, method-by-method explanations. Or may be just a series of blog posts ;)